app/abci/handlers.go

```diff
@@ -0,0 +1,270 @@
+package abci
+
+import (
+	"encoding/json"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+
+	addresscodec "cosmossdk.io/core/address"
+	"cosmossdk.io/log"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/sedaprotocol/seda-chain/app/utils"
+	batchingtypes "github.com/sedaprotocol/seda-chain/x/batching/types"
+)
+
+const (
+	// The block height difference from ExtendVote and VerifyVote to the
+	// corresponding batch's creation height.
+	voteExtensionOffset = -1
+	// The block height difference from PrepareProposal, ProcessProposal,
+	// and PreBlock to the corresponding batch's creation height.
+	proposalOffset = -2
+)
+
+type Handlers struct {
+	defaultPrepareProposal sdk.PrepareProposalHandler
+	defaultProcessProposal sdk.ProcessProposalHandler
+	batchingKeeper         BatchingKeeper
+	pubKeyKeeper           PubKeyKeeper
+	stakingKeeper          StakingKeeper
+	validatorAddressCodec  addresscodec.Codec
+	signer                 utils.SEDASigner
+	logger                 log.Logger
+}
+
+func NewHandlers(
+	dph *baseapp.DefaultProposalHandler,
+	bk BatchingKeeper,
+	pkk PubKeyKeeper,
+	sk StakingKeeper,
+	vac addresscodec.Codec,
+	logger log.Logger,
+) *Handlers {
+	return &Handlers{
+		defaultPrepareProposal: dph.PrepareProposalHandler(),
+		defaultProcessProposal: dph.ProcessProposalHandler(),
+		batchingKeeper:         bk,
+		pubKeyKeeper:           pkk,
+		stakingKeeper:          sk,
+		validatorAddressCodec:  vac,
+		logger:                 logger,
+	}
+}
+
+func (h *Handlers) SetSEDASigner(signer utils.SEDASigner) {
+	h.signer = signer
+}
+
+// ExtendVoteHandler handles the ExtendVote ABCI to sign a batch created
+// from the previous block.
+func (h *Handlers) ExtendVoteHandler() sdk.ExtendVoteHandler {
+	return func(ctx sdk.Context, _ *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
+		h.logger.Debug("start extend vote handler")
+
+		// Sign the batch created from the last block's end blocker.
+		batch, err := h.batchingKeeper.GetBatchForHeight(ctx, ctx.BlockHeight()+voteExtensionOffset)
+		if err != nil {
+			return nil, err
+		}
+		signature, err := h.signer.Sign(batch.BatchId, utils.Secp256k1)
+		if err != nil {
+			return nil, err
+		}
+
+		h.logger.Debug("submitting batch signature", "signature", signature)
+		return &abci.ResponseExtendVote{VoteExtension: signature}, nil
+	}
+}
+
+// VerifyVoteExtensionHandler handles the VerifyVoteExtension ABCI to
+// verify the batch signature included in the pre-commit vote against
+// the public key registered in the pubkey module.
+func (h *Handlers) VerifyVoteExtensionHandler() sdk.VerifyVoteExtensionHandler {
+	return func(ctx sdk.Context, req *abci.RequestVerifyVoteExtension) (*abci.ResponseVerifyVoteExtension, error) {
+		h.logger.Debug("start verify vote extension handler", "request", req)
+
+		if len(req.VoteExtension) != 64 {
```

Comment:
I think it makes sense to put a cap on the vote extension size, but if we make it the exact size of our signatures we're not forward compatible. Maybe it makes sense to reserve space for ~5 signatures? So we don't immediately run into problem when introducing new schemes.

Reply:
really good point. I added `const maxVoteExtensionLength = 64*5`. Also moved this check into `verifyBatchSignatures()`.